File: backend/app/brain/fuzzy.py
"""Fuzzy name matching — faithful port of lib/utils/fuzzy.ts.

Voice notes mangle names ("Arapham" for "Arfam"); the teammate lookup's
sound-alike fallback depends on these EXACT primitives. Divergence here means
the two brains resolve different people from the same voice note — port
byte-for-byte in spirit, extend both sides together.
"""

from __future__ import annotations

import re


def soundex(word: str) -> str:
    s = re.sub(r"[^A-Z]", "", word.upper())
    if not s:
        return ""

    def code(c: str) -> str:
        if c in "BFPV":
            return "1"
        if c in "CGJKQSXZ":
            return "2"
        if c in "DT":
            return "3"
        if c == "L":
            return "4"
        if c in "MN":
            return "5"
        if c == "R":
            return "6"
        return ""

    out = s[0]
    prev = code(s[0])
    for c in s[1:]:
        if len(out) >= 4:
            break
        d = code(c)
        if d and d != prev:
            out += d
        if c in "HW":
            continue
        prev = d
    return out.ljust(4, "0")


def levenshtein(a: str, b: str) -> int:
    s, t = a.lower(), b.lower()
    if s == t:
        return 0
    if not s:
        return len(t)
    if not t:
        return len(s)
    prev = list(range(len(t) + 1))
    for i in range(1, len(s) + 1):
        curr = [i]
        for j in range(1, len(t) + 1):
            curr.append(
                min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + (0 if s[i - 1] == t[j - 1] else 1))
            )
        prev = curr
    return prev[len(t)]


def name_matches_fuzzy(full_name: str, query: str) -> bool:
    q = query.strip()
    if len(q) < 2:
        return False
    q_code = soundex(q)
    max_edits = 2 if len(q) >= 5 else 1
    return any(
        soundex(tok) == q_code or levenshtein(tok, q) <= max_edits
        for tok in re.split(r"\s+", full_name)
        if len(tok) >= 2
    )

File: src/lib/utils/fuzzy.ts
// THE fuzzy name-matching utility — voice-transcription artifact resolution.
//
// Voice notes mangle names: Deepgram hears "Arapham" for "Arfam", "Evun" for
// "Evan" (real transcripts, 2026-08). A substring lookup finds nothing for
// these, so name resolution needs a SOUND-ALIKE fallback. Two pure primitives:
//
//   soundex()      — the classic phonetic code. "Arfam" and "Arapham" both
//                    encode A615, "Evan" and "Evun" both E150 — exactly the
//                    consonant-skeleton class of error STT produces.
//   levenshtein()  — edit distance, catches short typo-class drift the
//                    phonetic code misses ("Pawani" / "Pavani").
//
// nameMatchesFuzzy() combines them per name TOKEN (first/last name separately)
// so "Arapham" matches "Arfam Khan" via its first token.
//
// Pure functions, no I/O — usable in any layer. Consumers: the staff teammate
// lookup fallback (profiles-service searchTeammatesForElaya). R-01: never
// re-implement phonetic or edit-distance matching elsewhere — extend here.

/** American Soundex. Returns e.g. "A615". Empty input → "". */
export function soundex(word: string): string {
  const s = word.toUpperCase().replace(/[^A-Z]/g, "");
  if (s.length === 0) return "";
  const code = (c: string): string =>
    "BFPV".includes(c) ? "1"
    : "CGJKQSXZ".includes(c) ? "2"
    : "DT".includes(c) ? "3"
    : c === "L" ? "4"
    : "MN".includes(c) ? "5"
    : c === "R" ? "6"
    : ""; // vowels + H/W/Y drop
  let out = s[0];
  let prev = code(s[0]);
  for (let i = 1; i < s.length && out.length < 4; i++) {
    const c = s[i];
    const d = code(c);
    // H/W do not separate same-coded consonants; vowels do.
    if (d !== "" && d !== prev) out += d;
    if (c === "H" || c === "W") continue;
    prev = d;
  }
  return out.padEnd(4, "0");
}

/** Classic Levenshtein edit distance (iterative two-row). */
export function levenshtein(a: string, b: string): number {
  const s = a.toLowerCase();
  const t = b.toLowerCase();
  if (s === t) return 0;
  if (s.length === 0) return t.length;
  if (t.length === 0) return s.length;
  let prev = Array.from({ length: t.length + 1 }, (_, i) => i);
  for (let i = 1; i <= s.length; i++) {
    const curr = [i];
    for (let j = 1; j <= t.length; j++) {
      curr[j] = Math.min(
        prev[j] + 1,
        curr[j - 1] + 1,
        prev[j - 1] + (s[i - 1] === t[j - 1] ? 0 : 1),
      );
    }
    prev = curr;
  }
  return prev[t.length];
}

/**
 * Does `query` (one spoken/typed name, possibly mangled) plausibly refer to
 * `fullName`? True when ANY name token matches phonetically (same Soundex) or
 * within a small edit distance (≤1 for short tokens, ≤2 for 5+ chars).
 */
export function nameMatchesFuzzy(fullName: string, query: string): boolean {
  const q = query.trim();
  if (q.length < 2) return false;
  const qCode = soundex(q);
  const maxEdits = q.length >= 5 ? 2 : 1;
  return fullName
    .split(/\s+/)
    .filter((t) => t.length >= 2)
    .some((token) => soundex(token) === qCode || levenshtein(token, q) <= maxEdits);
}
